File: 0141-linked-list-cycle/solution.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public boolean hasCycle(ListNode head) {
        // Slow and fast pointer. Fast goes twice as fast as slow
        // When fast == slow, reset slow to start of list
        // Make fast go at same speed of slow
        // Where they meet again is where the loop starts

        if(head == null) return false;
        if(head.next == null) return false;
        else if (head.next.next == null) return false;

        ListNode slow = head;
        ListNode fast = head;

        do{
            if(slow == null || fast == null) return false;
            else if (slow.next == null || fast.next == null) return false;
            else if (slow.next.next == null || fast.next.next == null) return false;

            slow = slow.next;
            fast = fast.next.next;
        } while(slow != fast);

        if(slow.val == fast.val){
            return true;
        }

    return false;
    }
}

File: 0141-linked-list-cycle/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        # Tortoise and hare algorithm 
       
        if head is None or head.next is None:
            return False

        # 2 Pointers, 1 fast and one slow. If fast reaches None, there's no cycle
        # If they meet, there is a cycle
        # To find start of cycle (pos), one they meet, move fast back to head
        # Then increment boht pointers by 1 until they meet, where they meet is pos
        fast = head
        slow = head

        while(fast is not None):
            if fast.next is None:
                return False
            fast = fast.next.next
            slow = slow.next

            if fast is slow:
                return True
                
        return False


